apps/web/src/components/AssessmentStepTracker/utils.ts

```diff
@@ -3,17 +3,142 @@ import CheckCircleIcon from "@heroicons/react/20/solid/CheckCircleIcon";
 import ExclamationCircleIcon from "@heroicons/react/20/solid/ExclamationCircleIcon";
 import XCircleIcon from "@heroicons/react/20/solid/XCircleIcon";
 import PauseCircleIcon from "@heroicons/react/24/solid/PauseCircleIcon";
+import omit from "lodash/omit";
 
 import { IconType } from "@gc-digital-talent/ui";
-
 import {
+  PoolCandidate,
   ArmedForcesStatus,
   AssessmentDecision,
   AssessmentResult,
+  AssessmentResultType,
   Maybe,
-} from "~/api/generated";
+  PoolSkillType,
+} from "@gc-digital-talent/graphql";
+
 import { NO_DECISION, NullableDecision } from "~/utils/assessmentResults";
 
+export type CandidateAssessmentResult = {
+  poolCandidate: PoolCandidate;
+  decision: NullableDecision;
+  results: AssessmentResult[];
+};
+
+/**
+ * Derives the current decision of all assessments
+ * results for a specific set (attributed to a candidate)
+ *
+ * NOTE:
+ *
+ * Essential requirements = essential skills AND education requirements
+ * Optional requirements = asset skills
+ * Assessment Step Status =
+ *  IF any essential requirements unsuccessful THEN "unsuccessful"
+ *  IF any requirements unassessed THEN "to assess"
+ *  IF essential requirements include at least one hold THEN "hold"
+ *  IF all essential requirements successful THEN "success"
+ *
+ * @param results
+ * @returns NullableDecision
+ */
+const deriveResultsDecision = (
```

Comment:
When deriving the decision, you also have to take into account some details from the AssessmentStep itself. The current function checks the status of existing results, but I don't see it checking if the AssessmentStep requires skills for which there exists no Result at all.

Reply:
Hmm, okay I'll take a look. I assumed the fallback of `NO_DECISION` would be the catch all for missing assessment results though 🤔 Since, if a result does not exist it should fallback to "to assess", right?

Reply:
Falling back on NO_DECISION is correct. I'm worried about the situation where a user only has a single Result, and its successful, but has no results at all related to other essential skills.

Reply:
okay, that makes sense 👌 Thanks

Reply:
My suggestion for this function would be to:
1. assemble the list of requirements from the Assessment Step (essential skills and possibly education)
2. iterate through the requirements
  a) try to find a matching Result
    - if matching result is UNSUCCESSFUL, return UNSUCCESSFUL
    - if matching result is ON HOLD then set a HAS_ON_HOLD flag to true and continue
    - if matching result is TO ASSESS then set HAS_TO_ASSESS flag to true and continue
    - if matching result cannot be found then set HAS_TO_ASSESS flag to true and continue
    - if matching result is SUCCESSFUL then continue
3. Return according to flags:
  - If HAS_ON_HOLD return ON_HOLD
  - else if HAS_TO_ASSESS return TO_ASSESS
  - else return SUCCESSFUL

NOTE: you could replace the HAS_TO_ASSESS flag with a ALL_SUCCESSFUL flag that gets switched from true to false. Also, I'm not 100% certain about how ON_HOLD interacts with TO_ASSESS and SUCCESSFUL, now that I think about it... which takes priority? 🤔

Reply:
Went to look at #8644 to answer my own question. TO ASSESS takes priority over ON_HOLD:
```
3. Return according to flags:
  - If HAS_TO_ASSESS return TO_ASSESS
  - else if HAS_ON_HOLD return ON_HOLD
  - else return SUCCESSFUL
```  
If a user is ON_HOLD or SUCCESSFUL in one column, they should appear in the next column as well.

Reply:
okay, thank-you so much for all that detail! It really helped a lot 😄 

I did a major refactor of the functionality and it should be fixed in [refactor grouping of candidates](https://github.com/GCTC-NTGC/gc-digital-talent/pull/9022/commits/5df5503138a93af178dbd9956b446bd111f13323).

It is quite messy so I'd gladly take any suggestions 😅